## *Jeu de données IRVE*
-------------------
# Analyse des relations entre champs (intégrité des données)
------------------------------
## Contexte
- clarification du rôle des modèles de données dans les jeux de données (cf mise à jour récente des [guides data.gouv](https://guides.etalab.gouv.fr/qualite/documenter-les-donnees/))
- intégration d'une propriété "relationship" dans les schémas de données ([issue TableSchema](https://github.com/frictionlessdata/specs/issues/803) en cours de validation)
- création d'outils de contrôle des relations entre champs des jeux de données tabulaires (cf usage ci-dessous)

## Objectifs
- valider sur un cas réel l'utilisation d'un modèle de données en complément d'un schéma de données
- identifier les apports que pourraient avoir les contrôles de validation des relations entre champs

## Résultats
- les principaux résultats sont regroupés sur un [tableau de bord](https://nbviewer.org/github/loco-philippe/Environmental-Sensing/blob/main/python/Validation/irve/Analyse/IRVE_indicateurs.ipynb) 
- la formalisation d'un modèle de données facilite la compréhension des données et des relations entre champs
- l'outil de contrôle permet d'améliorer significativement la qualité des données par l'identification d'incohérences de relations
- l'identification des incohérences permet de trouver des stratégies de réduction des écarts (dans l'exemple ci-dessous, on passe 36% d'écart à 3,8 %)  
- l'analyse des données permet de (re)construire le modèle de données qui minimise les incohérences
- les incohérences détectées sur le jeu de données IRVE restent faibles (inférieures à 5 % des point de charge documentés - voir chapitre 4)

## Suite à donner
- Mettre à jour, valider et publier le modèle de données IRVE
- Définir les contrôles supplémentaires à intégrer pour toutes nouvelles données ainsi que pour le jeu complet
- Mettre en oeuvre les outils de contrôle

## Evolutions possibles 
- Ajouter dans les guides d'Etalab un guide pour les modèles de données 
- Intégrer dans les schémas de données la propriété "relationship" en cours de validation,
- Définir un indicateur qui mesure l'écart (existant / attendu) des relations entre champs

## Sommaire
*(liens actifs sur jupyter Notebook ou Nbviewer)*
- [1 - modèle de données](#1---modèle-de-données)
- [2 - Initialisation](#2---Initialisation)
- [3 - Séparation des pdc itinerance et hors itinerance](#3---Séparation-des-pdc-itinerance-et-hors-itinerance)
- [4 - Bilan initial intégrité](#4---Bilan-initial-intégrité)
- [5 - Séparation doublons pdc - date de maj](#5---Séparation-doublons-pdc---date-de-maj)
- [6 - Séparation doublons station - date de maj](#6---Séparation-doublons-station---date-de-maj)
- [7 - Synthèse](#7---Synthèse)
- [8 - Exemples d'erreurs résiduelles](#8---Exemples-d\'erreurs-résiduelles)

Ce Notebook peut être consulté sur [nbviewer](http://nbviewer.org/github/loco-philippe/Environmental-Sensing/tree/main/python/Validation/irve/Analyse)

données utilisées : https://www.data.gouv.fr/fr/datasets/fichier-consolide-des-bornes-de-recharge-pour-vehicules-electriques/    
fichier : "*consolidation-etalab-schema-irve-statique-v-2.2.0-2023xxxx.csv*"

-----------------------
# 1 - modèle de données

Le modèle de données proposé ci-dessous est construit sur la base du schéma de données mis à disposition et du contenu du jeu de données.    
Il est à consolider en fonction de l'expertise des concepteurs et réutilisateurs (voir [guide méthodologique](https://github.com/loco-philippe/Environmental-Sensing/blob/main/property_relationship/FR_methodology.ipynb)).

*Notation:*
- *M : Mandatory - documentation obligatoire*
- *PK : Primary Key - identifiant unique de l'entité*
- *Root : champ fictif associé à une ligne du tableau*

In [1]:
from base64 import b64encode
from IPython.display import Image, display
with open('IRVE_modele.txt', 'r', encoding="utf-8") as f:
    modele = f.read()
display(Image(url="https://mermaid.ink/img/" + b64encode(modele.encode("ascii")).decode("ascii")))

------
# 2 - Initialisation
## initialisation logicielle

In [2]:
from datetime import datetime
import json
from observation import Sdataset, Analysis
import pandas as pd

def analyse_integrite(data, fields, pr=True):
    '''analyse les relations du DataFrame définies dans le dict fields et retourne un dict avec pour chaque contrôle la
    liste des index ko. Les résultats des contrôles sont également ajoutés sous forme de champs booléens à data'''
    analyse = Analysis(data)
    dic_res = analyse.check_relationship(fields)
    data['ok'] = True
    for name, lis in dic_res.items():
        data[name] = True
        data.loc[lis, name] = False
        data['ok'] = data['ok'] & data[name]
        if pr:
            print('{:<50} {:>5}'.format(name, len(data) - data[name].sum()))
    return dic_res

def add_nbre_pdc(data, name='ok', pr=True):
    '''ajoute un champ avec le nombre de pdc calculés et le compare au champ 'nbre_pdc' '''
    data['nb_pdc_calc'] = data.groupby('id_station_itinerance')['index'].transform('count')
    data['nbre_pdc_ok'] = data['nb_pdc_calc'] == data['nbre_pdc']    
    data['calc-nbre'] = data['nb_pdc_calc'] - data['nbre_pdc']
    #data[name] = data[name] & data['nbre_pdc_ok']
    if pr:
        ecart_nbre = len(data) - data['nbre_pdc_ok'].sum()
        print('{:<20} {:>5}'.format('nbre_pdc_ko', ecart_nbre))
        print(round(ecart_nbre / len(data) * 100), ' %')

## initialisation des données
- lecture du fichier issu de l'api

In [3]:
dates = ['2023-03-03', '2023-04-17', '2023-05-02', '2023-05-24', '2023-06-17', '2023-07-04', '2023-07-23', '2023-08-08']
date= dates[-1]
log = {'date_irve': date, 
    'file': 'consolidation-etalab-schema-irve-statique-v-2.2.0-'+date[:4]+date[5:7]+date[8:]+'.csv',
#    'chemin': 'https://raw.githubusercontent.com/loco-philippe/Environmental-Sensing/gen-ntv/python/Validation/irve/Analyse/'}
#log = {'date_irve': '2022-06-06', 'file': 'consolidation-etalab-schema-irve-v-2.0.2-20220606-propre2.csv',
    'chemin': 'D:\\philippe\\python ESstandard\\Environmental-Sensing\\python\\Validation\\irve\\Analyse\\'}
irve = pd.read_csv(log['chemin'] + log['file'], sep=',', low_memory=False)
log['len_irve'] = len(irve)
print('nombre de lignes : ', log['len_irve']) 

nombre de lignes :  41232


## schéma de données
Le schéma de données restreint à la propriété 'relationship' et construit à partir du modèle de données est le suivants :

In [4]:
# complément à inclure dans le schéma de données
fields = [
 # relation unicité des pdl
 { "name": "index",
   "relationship" : { "parent" : "id_pdc_itinerance", "link" : "coupled" }},   
 # relations inter entités
 { "name": "contact_operateur",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nom_enseigne",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "coordonneesXY",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "id_station_itinerance",
   "relationship" : { "parent" : "id_pdc_itinerance",     "link" : "derived" }},
 # relations intra entité - station
 { "name": "nom_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "implantation_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 #{ "name": "date_maj",
 #  "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nbre_pdc",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "condition_acces",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "horaires",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "station_deux_roues",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 # relations intra entité - localisation
 { "name": "adresse_station",
   "relationship" : { "parent" : "coordonneesXY",         "link" : "derived" }} ]

# liste des champs liés à un controle (relations) et obligatoires (mandatory)
relations = ['index', 'contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 
             'nom_station', 'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 
             'id_pdc_itinerance', 'date_maj', 'last_modified']
mandatory = ['contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 'nom_station',
             'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 'id_pdc_itinerance', 
             'puissance_nominale', 'prise_type_ef', 'prise_type_2', 'prise_type_combo_ccs', 'prise_type_chademo', 
             'prise_type_autre', 'paiement_acte', 'paiement_autre', 'reservation',  'accessibilite_pmr', 'restriction_gabarit', 
             'date_maj', 'last_modified']
controles = ['index - id_pdc_itinerance', 'contact_operateur - id_station_itinerance', 'nom_enseigne - id_station_itinerance',
             'coordonneesXY - id_station_itinerance', 'id_station_itinerance - id_pdc_itinerance',
             'nom_station - id_station_itinerance', 'implantation_station - id_station_itinerance',
             'nbre_pdc - id_station_itinerance', 'condition_acces - id_station_itinerance', 'horaires - id_station_itinerance',
             'station_deux_roues - id_station_itinerance', 'adresse_station - coordonneesXY']

-----------------------
## 3 - Séparation des pdc itinerance et hors itinerance
- de l'ordre de 1 % des points de charge sont hors itinerance

In [5]:
# séparation des données
data = irve
data[['id_station_itinerance','id_pdc_itinerance']] = data[['id_station_itinerance','id_pdc_itinerance']].astype('string')
#data[['last_modified','date_maj']] = data[['last_modified','date_maj']].astype('datetime64')
data['non_concerne'] = data['id_station_itinerance'].str.contains('oncern') | data['id_pdc_itinerance'].str.contains('oncern')

non_concerne = data[data['non_concerne']].reset_index()['index']
itinerance = data[~data['non_concerne']].reset_index()
itinerance_init = itinerance.loc[:, relations]
log['pdc_hors_itinerance'] = len(non_concerne)
log['pdc_en_itinerance'] = len(itinerance)
print('nombre de pdc hors itinerance : ', log['pdc_hors_itinerance'])
print('nombre de pdc en itinerance   : ', log['pdc_en_itinerance'])

nombre de pdc hors itinerance :  2318
nombre de pdc en itinerance   :  38914


In [6]:
# arborescence des champs et nombre de valeurs différentes
il = Sdataset(itinerance_init)
print(il.tree())

-1: root-derived (38914)
   0 : index (38914)
      1 : contact_operateur (128)
      2 : nom_enseigne (1007)
      3 : coordonneesXY (12670)
      4 : adresse_station (12176)
      5 : id_station_itinerance (17197)
      6 : nom_station (12775)
      7 : implantation_station (5)
      8 : nbre_pdc (33)
      9 : condition_acces (2)
      10: horaires (299)
      11: station_deux_roues (6)
      12: id_pdc_itinerance (36957)
      13: date_maj (600)
      14: last_modified (494)


-----------------------
## 4 - Bilan initial intégrité
- plus de 50 % des lignes présentent un défaut d'intégrité

In [7]:
# séparation données bonnes (itinerance_ok_1) et données résiduelles (itinerance_1)
res = analyse_integrite(itinerance_init, fields)
itinerance_ok_1 = itinerance_init.loc[itinerance_init.ok, relations].reset_index(drop=True)
itinerance_1 = itinerance_init.loc[~itinerance_init.ok, relations].reset_index(drop=True)
itinerance_init = itinerance_init.loc[:, relations]
log['init_ok'] = len(itinerance_ok_1)
log['init_ko'] = len(itinerance_1)
print("\nnombre d'enregistrements sans erreurs : ", log['init_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['init_ko'])
print("taux d'erreur : ", round(log['init_ko'] / log['pdc_en_itinerance'] * 100), ' %')

index - id_pdc_itinerance                           3868
contact_operateur - id_station_itinerance            997
nom_enseigne - id_station_itinerance                2282
coordonneesXY - id_station_itinerance               1882
id_station_itinerance - id_pdc_itinerance            697
nom_station - id_station_itinerance                 1579
implantation_station - id_station_itinerance        1332
nbre_pdc - id_station_itinerance                     823
condition_acces - id_station_itinerance               31
horaires - id_station_itinerance                      97
station_deux_roues - id_station_itinerance          1409
adresse_station - coordonneesXY                     1102

nombre d'enregistrements sans erreurs :  32534
nombre d'enregistrements avec au moins une erreur :  6380
taux d'erreur :  16  %


-----------------------
## 5 - Séparation doublons pdc - date de maj
- plus de la moitié des pdc en erreur sont liées aux doublons de pdc
- la suppression des doublons permet de réduire de plus de 80% le nombre de lignes erronnées

In [8]:
# séparation doublons pdc (doublons_pdc) et données résiduelles (itinerance_2)
itinerance_1['doublons_pdc'] = itinerance_1.sort_values(by=['date_maj', 'last_modified']).duplicated('id_pdc_itinerance', keep='last')

doublons_pdc = itinerance_1[itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)['index']
#doublons_pdc = itinerance_1[itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)
itinerance_2 = itinerance_1[~itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)
itinerance_1 = itinerance_1.loc[:, relations]
log['doublons_pdc'] = len(doublons_pdc)
log['sans_doublons_pdc'] = len(itinerance_2)
print('nombre de doublons pdc : ', log['doublons_pdc'], ' soit ', round(log['doublons_pdc']/log['init_ko'] * 100), ' %')
print('nombre de pdc sans doublon   : ', log['sans_doublons_pdc'])

nombre de doublons pdc :  1957  soit  31  %
nombre de pdc sans doublon   :  4423


In [9]:
# séparation données bonnes (itinerance_ok_3) et données résiduelles (itinerance_3)
res = analyse_integrite(itinerance_2, fields)
itinerance_ok_3 = itinerance_2.loc[itinerance_2.ok, relations].reset_index(drop=True)
itinerance_3 = itinerance_2.loc[~itinerance_2.ok, relations].reset_index(drop=True)
itinerance_2 = itinerance_2.loc[:, relations]
log['etape3_ok'] = len(itinerance_ok_3)
log['etape3_ko'] = len(itinerance_3)
print("\nnombre d'enregistrements sans erreurs : ", log['etape3_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['etape3_ko'])

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance             67
nom_enseigne - id_station_itinerance                 337
coordonneesXY - id_station_itinerance                920
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                  477
implantation_station - id_station_itinerance         287
nbre_pdc - id_station_itinerance                     789
condition_acces - id_station_itinerance               13
horaires - id_station_itinerance                      56
station_deux_roues - id_station_itinerance           513
adresse_station - coordonneesXY                      356

nombre d'enregistrements sans erreurs :  1903
nombre d'enregistrements avec au moins une erreur :  2520


-----------------------
## 6 - Séparation doublons station - date de maj
- plus de 15% des erreurs résiduelles sont liées au mélange d'anciens et de nouveaux pdc
- la suppression des anciens pdc permet de réduire de plus de 20% le nombre de lignes erronnées
- les dernières erreurs correspondent à des stations associées à 33 opérateurs et sont liées à des causes multiples

In [10]:
# séparation doublons stations (doublons_stat_maj) et données résiduelles (itinerance_4)
itinerance_3['stat_maj'] = itinerance_3.id_station_itinerance + itinerance_3.date_maj
stat_maj_unique = itinerance_3.sort_values(by='stat_maj').drop_duplicates('id_station_itinerance', keep='last')
itinerance_3['last_stat_maj'] = itinerance_3['stat_maj'].isin(stat_maj_unique['stat_maj'])

doublons_stat_maj = itinerance_3[~itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)['index']
itinerance_4 = itinerance_3[itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)
itinerance_3 = itinerance_3.loc[:, relations]
log['doublons_station'] = len(doublons_stat_maj)
log['sans_doublons_station'] = len(itinerance_4)
print('nombre de doublons stations : ', log['doublons_station'], ' soit ', 
      round(log['doublons_station']/log['etape3_ko'] * 100), ' %')
print('nombre de pdc sans doublon   : ', log['sans_doublons_station'])

nombre de doublons stations :  479  soit  19  %
nombre de pdc sans doublon   :  2041


In [11]:
# séparation données bonnes (itinerance_ok_5) et données résiduelles (itinerance_5 / itinerance_5_full)
res = analyse_integrite(itinerance_4, fields)
itinerance_ok_5 = itinerance_4.loc[itinerance_4.ok, relations].reset_index(drop=True)
itinerance_5_full = itinerance_4.loc[~itinerance_4.ok].reset_index(drop=True)
itinerance_5 = itinerance_5_full.loc[:, relations]
#itinerance_4 = itinerance_4.loc[:, relations]
log['etape5_ok'] = len(itinerance_ok_5)
log['etape5_ko'] = len(itinerance_5)
print("\nnombre d'enregistrements sans erreurs : ", log['etape5_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['etape5_ko'], ' soit ', 
      round(log['etape5_ko']/log['etape3_ko'] * 100), ' %')

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance              2
nom_enseigne - id_station_itinerance                 263
coordonneesXY - id_station_itinerance                377
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                  195
implantation_station - id_station_itinerance          52
nbre_pdc - id_station_itinerance                     539
condition_acces - id_station_itinerance                8
horaires - id_station_itinerance                      52
station_deux_roues - id_station_itinerance           448
adresse_station - coordonneesXY                      319

nombre d'enregistrements sans erreurs :  99
nombre d'enregistrements avec au moins une erreur :  1942  soit  77  %


In [12]:
# structure des données résiduelles et nombre de valeurs
il = Sdataset(itinerance_5)
print(il.tree())

-1: root-derived (1942)
   0 : index (1942)
      1 : contact_operateur (31)
      2 : nom_enseigne (338)
      3 : coordonneesXY (479)
      4 : adresse_station (515)
         14: last_modified (60)
      5 : id_station_itinerance (523)
         13: date_maj (75)
      6 : nom_station (472)
      7 : implantation_station (5)
      8 : nbre_pdc (17)
      9 : condition_acces (2)
      10: horaires (69)
      11: station_deux_roues (5)
      12: id_pdc_itinerance (1942)


In [13]:
# nb maxi d'erreurs
itinerance_4['somme'] = 0
for name in res.keys():
    itinerance_4['somme'] += 1 - itinerance_4[name]
erreurs = max(itinerance_4['somme'])
maxi = itinerance_4[itinerance_4.somme >= erreurs]
print("nombre d'enregistrements avec ", erreurs, " erreurs : ", len(maxi))

nombre d'enregistrements avec  4  erreurs :  4


-----------------------
## 7 - Synthèse


### fichiers
Génération des fichiers intégrant les défauts d'intégrité :
- fichier csv des lignes résiduelles à traiter (IRVE_itinerance_residuel)
- fichier csv des données itinerance avec indicateur des données à corriger ou à ignorer (IRVE_itinerance_complet)
- fichier csv des données itinerance valides (IRVE_itinerance_valide)
- fichier csv des doublons (IRVE_itinerance_doublons)

In [14]:
# consolidation des données
itinerance['doublons_stat_maj'] = itinerance['index'].isin(doublons_stat_maj)
itinerance['doublons_pdc'] = itinerance['index'].isin(doublons_pdc)
itinerance['lignes_a_corriger'] = itinerance['index'].isin(itinerance_5['index'])
itinerance['doublons_a_supprimer'] = itinerance['doublons_stat_maj'] | itinerance['doublons_pdc']
itinerance['lignes_ko'] = itinerance['doublons_a_supprimer'] | itinerance['lignes_a_corriger']
print('total des lignes à corriger : ', itinerance['lignes_a_corriger'].sum())
itinerance_doublons = itinerance[itinerance['doublons_a_supprimer']].reset_index(drop=True)
print('total des doublons à supprimer : ', len(itinerance_doublons))
itinerance_ok = itinerance[~itinerance['lignes_ko']].reset_index(drop=True)
print('nombre de pdc avec controles ok : ', len(itinerance_ok))

total des lignes à corriger :  1942
total des doublons à supprimer :  2436
nombre de pdc avec controles ok :  34536


In [15]:
#génération des fichiers
extension = log['date_irve'] +'.csv'
itinerance_5_full.to_csv('IRVE_itinerance_residuel' + extension)
itinerance.to_csv('IRVE_itinerance_complet' + extension)
itinerance_ok.to_csv('IRVE_itinerance_valide' + extension)
itinerance_doublons.to_csv('IRVE_itinerance_doublons' + extension)
log['IRVE_itinerance_residuel' + extension] = len(itinerance_5_full)
log['IRVE_itinerance_complet' + extension] = len(itinerance)
log['IRVE_itinerance_valide' + extension] = len(itinerance_ok)
log['IRVE_itinerance_valide_stat' + extension] = len(itinerance_ok.drop_duplicates('id_station_itinerance', keep='last'))
log['IRVE_itinerance_doublons' + extension] = len(itinerance_doublons)

### vérification de l'intégrité

In [16]:
# vérification de l'absence d'erreurs
res = analyse_integrite(itinerance_ok.loc[:, relations], fields, pr=False)
log['bilan_erreurs'] = sum([len(controle) for controle in res.values()])
log['date'] = datetime.now().isoformat()
print('bilan intégrité :')
print('    erreurs : ', log['bilan_erreurs'])

bilan intégrité :
    erreurs :  0


In [17]:
# structure des données bonnes
il = Sdataset(itinerance_ok.loc[:, mandatory])
print(il.tree())

-1: root-derived (34536)
   11: id_pdc_itinerance (34536)
      4 : id_station_itinerance (16491)
         1 : nom_enseigne (675)
         2 : coordonneesXY (11705)
            0 : contact_operateur (111)
            3 : adresse_station (11088)
            8 : condition_acces (2)
         5 : nom_station (11807)
         6 : implantation_station (5)
         7 : nbre_pdc (33)
         9 : horaires (235)
         10: station_deux_roues (5)
         19: paiement_autre (6)
      12: puissance_nominale (72)
      13: prise_type_ef (8)
      14: prise_type_2 (7)
      15: prise_type_combo_ccs (7)
      16: prise_type_chademo (7)
      17: prise_type_autre (6)
      18: paiement_acte (8)
      20: reservation (6)
      21: accessibilite_pmr (4)
      22: restriction_gabarit (108)
      23: date_maj (437)
      24: last_modified (407)


### Indicateurs

In [18]:
# génération des indicateurs liés aux règles
regles = ['Pdc non unique', 'Station multi-operateurs', 'Station multi-enseignes', 'Station multi-localisations', 
          'Pdc multi-stations', 'station avec plusieurs noms', 'station multi-implantations', 
          'nombre de pdc par station incoherent', 'station multi-acces', 'station multi-horaires', 
          'acces deux-roues incoherent', 'localisation multi-adresses']
principal = [16, 17, 18, 19, 20]
secondaire = [21, 22, 23, 24, 25, 26, 27]
irve = itinerance_5_full # residuel
total = len(irve)
indic = {}

irve['principal'] = True
for ind in principal:
    irve['principal'] &= irve.iloc[:,ind]
    indic[regles[ind-16]] = int(total - irve.iloc[:,ind].sum())
irve['secondaire'] = True
for ind in secondaire:
    irve['secondaire'] &= irve.iloc[:,ind]
    indic[regles[ind-16]] = int(total - irve.iloc[:,ind].sum())
irve['secondaire'] |= (~irve['principal'] & ~irve['secondaire'])
irve['verif'] = irve['principal'] & irve['secondaire']
indic['principal pdc'] = int(total - irve['principal'].sum())
indic['secondaire pdc'] = int(total - irve['secondaire'].sum())

irve_p = irve[~irve['principal']].drop_duplicates('id_station_itinerance').reset_index(drop=True)
irve_s = irve[~irve['secondaire']].drop_duplicates('id_station_itinerance').reset_index(drop=True)
indic['principal stat'] = len(irve_p)
indic['secondaire stat'] = len(irve_s)



In [19]:
# ajout des principaux opérateurs contributeur de défauts
operateurs = list(itinerance_5.drop_duplicates('contact_operateur')['contact_operateur'])
erreurs_op = [len(itinerance_5.loc[itinerance_5.contact_operateur == operateur]) for operateur in operateurs]
err_op, oper = tuple(zip(*(sorted(zip(erreurs_op, operateurs), reverse=True))[:5]))
indic['operateurs'] = list(oper)
indic['erreurs_operateurs'] = list(err_op)

print(len(err_op), ' opérateurs représentent : ', round(sum(err_op) / sum(erreurs_op) * 100), ' % des écarts\n')
for err, op in zip(indic['erreurs_operateurs'], indic['operateurs']):
    print(op, err)

5  opérateurs représentent :  76  % des écarts

supervision-ev.france@totalenergies.com 506
assistance-commerciale@e-vadea.fr 321
pascal.lhermitte@siege27.fr 260
web@freshmile.com 207
contact@e55c.com 184


In [20]:
# ajout de l'indicateur d'écart entre nombre de pdc calculés et saisis
add_nbre_pdc(irve, pr=True)
indic['nb_pdc_calc-nbre_pdc'] = int(irve['calc-nbre'].sum())

nbre_pdc_ko            876
45  %


In [21]:
# stockage des indicateurs
log |= indic
with open('logfile.txt', 'a', encoding="utf-8") as f:
    f.write(json.dumps(log) + '\n')

-----------------------
## 8 - Exemples d'erreurs résiduelles 

### Erreurs multiples 
- exemple avec le nombre maximal d'erreurs (4) (2 stations soit 29 pdc avec : adresse, nom, nbre_pdc et deux-roues erronés)

In [22]:
maxi.loc[:, relations]

index contact_operateur          nom_enseigne          coordonneesXY  \
1493  34784  contact@e55c.com  ELECTRIC 55 CHARGING  [2.130358, 48.801996]   
1494  34785  contact@e55c.com  ELECTRIC 55 CHARGING  [2.130358, 48.801996]   
1689  35234  contact@e55c.com  ELECTRIC 55 CHARGING  [2.130358, 48.801996]   
1690  35235  contact@e55c.com  ELECTRIC 55 CHARGING  [2.130358, 48.801996]   

                    adresse_station id_station_itinerance  \
1493     AVENUE DE PARIS VERSAILLES           FR55CP78000   
1494     AVENUE DE PARIS VERSAILLES           FR55CP78000   
1689  6 AVENUE DE PARIS -VERSAILLES           FR55CP78000   
1690  6 AVENUE DE PARIS -VERSAILLES           FR55CP78000   

                        nom_station implantation_station  nbre_pdc  \
1493     AVENUE DE PARIS VERSAILLES               Voirie         4   
1494     AVENUE DE PARIS VERSAILLES               Voirie         4   
1689  6 AVENUE DE PARIS -VERSAILLES               Voirie         2   
1690  6 AVENUE DE PARIS -VERSAILLES               Voirie         2   

     condition_acces horaires station_deux_roues            id_pdc_itinerance  \
1493   Accès réservé     24/7              FALSE  FR55CE780004880188021299351   
1494   Accès réservé     24/7              FALSE  FR55CE780004880188021299331   
1689   Accès réservé     24/7              FALSE      FR55CEFR78000AVDEPAR1S0   
1690   Accès réservé     24/7              FALSE      FR55CEFR78000AVDEPAR1S1   

        date_maj                     last_modified  
1493  2022-11-03  2023-08-08T07:16:44.947000+00:00  
1494  2022-11-03  2023-08-08T07:16:44.947000+00:00  
1689  2022-11-03  2023-08-08T07:16:44.947000+00:00  
1690  2022-11-03  2023-08-08T07:16:44.947000+00:00

### Cohérence implantation_station - id_station
- 64 pdc sont liés à une erreur de choix d'implantation (3 stations, 1 opérateur)

In [23]:
itinerance_4.loc[~itinerance_4['implantation_station - id_station_itinerance'], relations]

index contact_operateur                        nom_enseigne  \
45     5970    info@allego.eu        Geant Casino Aix en Provence   
46     5971    info@allego.eu        Geant Casino Aix en Provence   
47     5972    info@allego.eu        Geant Casino Aix en Provence   
48     5974    info@allego.eu            Geant Casino Saint Louis   
49     5975    info@allego.eu            Geant Casino Saint Louis   
50     5976    info@allego.eu            Geant Casino Saint Louis   
51     6056    info@allego.eu     Allego Geant Casino Montpellier   
52     6057    info@allego.eu     Allego Geant Casino Montpellier   
53     6558    info@allego.eu         Allego Geant Casino Avignon   
54     6563    info@allego.eu         Allego Geant Casino Avignon   
55     6564    info@allego.eu         Allego Geant Casino Avignon   
56     6565    info@allego.eu         Allego Geant Casino Avignon   
57     6600    info@allego.eu        Allego Geant Casino Lanester   
58     6601    info@allego.eu        Allego Geant Casino Lanester   
59     6602    info@allego.eu        Allego Geant Casino Lanester   
60     6603    info@allego.eu        Allego Geant Casino Lanester   
61     6711    info@allego.eu      Allego Geant Casino Champniers   
62     6712    info@allego.eu      Allego Geant Casino Champniers   
63     6713    info@allego.eu      Allego Geant Casino Champniers   
64     6714    info@allego.eu      Allego Geant Casino Champniers   
65     6720    info@allego.eu  Allego Geant Casino Tours La Riche   
66     6721    info@allego.eu  Allego Geant Casino Tours La Riche   
67     6722    info@allego.eu  Allego Geant Casino Tours La Riche   
68     6723    info@allego.eu  Allego Geant Casino Tours La Riche   
69     6728    info@allego.eu      Allego Geant Casino Fenouillet   
70     6729    info@allego.eu      Allego Geant Casino Fenouillet   
71     6752    info@allego.eu           Allego Geant Casino Nîmes   
72     6754    info@allego.eu           Allego Geant Casino Nîmes   
73     6755    info@allego.eu           Allego Geant Casino Nîmes   
74     6756    info@allego.eu           Allego Geant Casino Nîmes   
75     6763    info@allego.eu         Allego Geant Casino Limoges   
76     6784    info@allego.eu         Allego Geant Casino Limoges   
77     6869    info@allego.eu         Allego Geant Casino Limoges   
78     6870    info@allego.eu         Allego Geant Casino Limoges   
79     6949    info@allego.eu         Allego Geant Casino Valence   
80     7120    info@allego.eu                        Allego Arcos   
81     7121    info@allego.eu                        Allego Arcos   
82     7137    info@allego.eu          Allego Geant Casino Frejus   
83     7138    info@allego.eu          Allego Geant Casino Frejus   
84     7151    info@allego.eu         Allego Geant Casino Valence   
85     7152    info@allego.eu         Allego Geant Casino Valence   
86     7153    info@allego.eu         Allego Geant Casino Valence   
87     7154    info@allego.eu          Allego Geant Casino Frejus   
88     7155    info@allego.eu          Allego Geant Casino Frejus   
89     7156    info@allego.eu          Allego Geant Casino Frejus   
90     7166    info@allego.eu          Allego Geant Casino Frejus   
91     7172    info@allego.eu          Allego Geant Casino Frejus   
92     7173    info@allego.eu          Allego Geant Casino Frejus   
1509  34815  contact@e55c.com                ELECTRIC 55 CHARGING   
1510  34816  contact@e55c.com                ELECTRIC 55 CHARGING   
1511  34817  contact@e55c.com                ELECTRIC 55 CHARGING   
1512  34818  contact@e55c.com                ELECTRIC 55 CHARGING   

                  coordonneesXY  \
45    [5.41683791, 43.53314177]   
46    [5.41683791, 43.53314177]   
47    [5.41683791, 43.53314177]   
48        [7.557195, 47.583132]   
49        [7.557195, 47.583132]   
50        [7.557195, 47.583132]   
51         [3.889034, 43.58858]   
52         [3.889034, 43.58858]   
53     [4.83458514, 43.9278626]  

### Cohérence nom_station - id_station
- 328 pdc sont associés à une station avec un nom non cohérent

In [24]:
itinerance_4.loc[~itinerance_4['nom_station - id_station_itinerance'], relations]

index                        contact_operateur          nom_enseigne  \
771   22610                       bo_dgic@seolis.net             AlterBase   
772   22611                       bo_dgic@seolis.net             AlterBase   
1347  29333  supervision-ev.france@totalenergies.com          INTERPARKING   
1348  29334  supervision-ev.france@totalenergies.com          INTERPARKING   
1349  29335  supervision-ev.france@totalenergies.com          INTERPARKING   
...     ...                                      ...                   ...   
1690  35235                         contact@e55c.com  ELECTRIC 55 CHARGING   
1878  36740              pascal.lhermitte@siege27.fr               SIEGE50   
1892  36754              pascal.lhermitte@siege27.fr               SIEGE49   
1952  36815              pascal.lhermitte@siege27.fr               SIEGE87   
1953  36816              pascal.lhermitte@siege27.fr               SIEGE88   

                coordonneesXY                      adresse_station  \
771   [-0.4604029,46.3288736]           Place Chanzy - 79000 NIORT   
772   [-0.4604029,46.3288736]           Place Chanzy - 79000 NIORT   
1347    [7.265881, 43.695113]    1 Promenade des anglais 6000 NICE   
1348    [7.265881, 43.695113]    1 Promenade des anglais 6000 NICE   
1349    [7.265881, 43.695113]    1 Promenade des anglais 6000 NICE   
...                       ...                                  ...   
1690    [2.130358, 48.801996]        6 AVENUE DE PARIS -VERSAILLES   
1878    [1.221452, 49.197647]  Ecoparc satellite ZA1 Heudebouville   
1892    [1.221452, 49.197647]  Ecoparc satellite ZA1 Heudebouville   
1952    [1.224898, 49.195263]  Ecoparc satellite ZA2 Heudebouville   
1953    [1.224898, 49.195263]  Ecoparc satellite ZA2 Heudebouville   

           id_station_itinerance  \
771          FRSEOPAB79191P0031A   
772          FRSEOPAB79191P0031A   
1347                   FRIPKNRME   
1348                   FRIPKNRME   
1349                   FRIPKNRME   
...                          ...   
1690                 FR55CP78000   
1878  FRS27PHEUDEBOUVILLEECOPARC   
1892  FRS27PHEUDEBOUVILLEECOPARC   
1952  FRS27PHEUDEBOUVILLEECOPARC   
1953  FRS27PHEUDEBOUVILLEECOPARC   

                                            nom_station  \
771   Réseau AlterBase - Niort - Conseil Département...   
772   Réseau AlterBase - Niort - Conseil Département...   
1347                 INTERPARKING  - Nice Ruhl Méridien   
1348                 INTERPARKING  - Nice Ruhl Méridien   
1349                 INTERPARKING  - Nice Ruhl Méridien   
...                                                 ...   
1690                      6 AVENUE DE PARIS -VERSAILLES   
1878                                             900012   
1892                                             900012   
1952                                             900031   
1953                                             900031   

              implantation_station  nbre_pdc condition_acces horaires  \
771                 Parking public         2     Accès libre     24/7   
772                 Parking public         2     Accès libre     24/7   
1347  Parking privé à usage public        22     Accès libre     24/7   
1348  Parking privé à usage public        22     Accès libre     24/7   
1349  Parking privé à usage public        22     Accès libre     24/7   
...                            ...       ...             ...      ...   
1690                        Voirie         2   Accès réservé     24/7   
1878                        Voirie         2     Accès libre     24/7   
1892                        Voirie         2     Accès libre     24/7   
1952                        Voirie         2     Accès libre     24/7   
1953                        Voirie         2     Accès libre     24/7   

     station_deux_roues            id_pdc_itinerance    date_maj  \
771               false        FRSEOEAB79191P0031A11  2023-07-25   
772               false        FRSEOEAB79191P0031A12  2023-07-25   
1347 

### Cohérence adresse - coordonnées
- 440 pdc ont une adresse non cohérente avec les coordonnées géographiques

In [25]:
itinerance_4.loc[~itinerance_4['adresse_station - coordonneesXY'], relations]

index       contact_operateur               nom_enseigne  \
0      1354          sav@izivia.com                     SIGEIF   
1      1355          sav@izivia.com                     SIGEIF   
2      1358          sav@izivia.com                     SIGEIF   
3      1360          sav@izivia.com                     SIGEIF   
4      2981          sav@izivia.com                   SOREGIES   
...     ...                     ...                        ...   
2007  37573  contact@ubitricity.com  Ubitricity France Network   
2008  37658  contact@ubitricity.com  Ubitricity France Network   
2009  37659  contact@ubitricity.com  Ubitricity France Network   
2010  37660  contact@ubitricity.com  Ubitricity France Network   
2011  37661  contact@ubitricity.com  Ubitricity France Network   

              coordonneesXY  \
0        [2.27736, 48.9917]   
1        [2.27736, 48.9917]   
2        [2.27736, 48.9917]   
3        [2.27736, 48.9917]   
4       [0.30007, 46.54829]   
...                     ...   
2007  [0.223944, 49.564167]   
2008  [0.073833, 49.510417]   
2009  [0.073833, 49.510417]   
2010  [0.223944, 49.564167]   
2011  [0.223944, 49.564167]   

                                        adresse_station id_station_itinerance  \
0               RUE CRISTINO GARCIA PLACE ETIENNE DOLET       FRSIGPSIGE34212   
1               RUE CRISTINO GARCIA PLACE ETIENNE DOLET       FRSIGPSIGE34211   
2                                   RUE CRISTINO GARCIA       FRSIGPSIGE34411   
3                                   RUE CRISTINO GARCIA       FRSIGPSIGE34412   
4                                     RUE DE LÉCORCERIE       FRS86PSOREB1511   
...                                                 ...                   ...   
2007                   Rue de Chemin de Fer - Epouville        FRUBIE10062883   
2008  30 Rue Georges Boissaye du Bocage - Sainte-Adr...        FRUBIE10099360   
2009  30 Rue Georges Boissaye du Bocage - Sainte-Adr...        FRUBIE10095460   
2010                   Rue du Chemin de Fer - Épouville        FRUBIE10062826   
2011                   Rue du Chemin de Fer - Épouville        FRUBIE10062738   

                                            nom_station implantation_station  \
0     SIGEIF - 4 RUE CRISTINO - GARCIA PLACE ETIENNE...               Voirie   
1     SIGEIF - 4 RUE CRISTINO - GARCIA PLACE ETIENNE...               Voirie   
2     SIGEIF - 4 RUE CRISTINO GARCIA - PLACE ETIENNE...               Voirie   
3     SIGEIF - 4 RUE CRISTINO GARCIA - PLACE ETIENNE...               Voirie   
4                                      POITIERS SUD HPC               Voirie   
...                                                 ...                  ...   
2007                   Rue de Chemin de Fer - Epouville       Parking public   
2008  30 Rue Georges Boissaye du Bocage - Sainte-Adr...       Parking public   
2009  30 Rue Georges Boissaye du Bocage - Sainte-Adr...       Parking public   
2010                   Rue du Chemin de Fer - Épouville       Parking public   
2011                   Rue du Chemin de Fer - Épouville       Parking public   

      nbre_pdc condition_acces horaires station_deux_roues id_pdc_itinerance  \
0            2     Accès libre     24/7              false   FRSIGESIGE34212   
1            2     Accès libre     24/7              false   FRSIGESIGE34211   
2            2     Accès libre     24/7              false   FRSIGESIGE34411   
3            2     Accès libre     24/7              false   FRSIGESIGE34412   
4            1     Accès libre     24/7              false   FRS86ESOREB1511   
...        ...             ...      ...                ...               ...   
2007         1     Accès libre     24/7              FALSE    FRUBIE10062883   
2008         1     Accès libre     24/7              FALSE    FRUBIE10099360   
2009         1     Accès libre     24/7              FALSE    FRUBIE10095460   
2010         1     Accès libre     24/7              FALSE    FRUBIE10062826   
2011         1     Accès l

- exemple : une station avec plusieurs adresses -> erreur de saisie ?

In [26]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[2.87930851314442, 48.94679007929618]', relations]

index                        contact_operateur   nom_enseigne  \
1239  28455  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1240  28456  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1241  28457  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1242  28469  supervision-ev.france@totalenergies.com  RIESTER MEAUX   

                              coordonneesXY  \
1239  [2.87930851314442, 48.94679007929618]   
1240  [2.87930851314442, 48.94679007929618]   
1241  [2.87930851314442, 48.94679007929618]   
1242  [2.87930851314442, 48.94679007929618]   

                            adresse_station id_station_itinerance  \
1239  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1240  51 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1241  50 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1242  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   

        nom_station          implantation_station  nbre_pdc condition_acces  \
1239  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1240  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1241  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1242  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   

                                               horaires station_deux_roues  \
1239  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1240  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1241  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1242  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   

     id_pdc_itinerance    date_maj                     last_modified  
1239      FRTCBE003958  2023-01-27  2023-08-08T07:17:25.387000+00:00  
1240      FRTCBE003955  2023-01-27  2023-08-08T07:17:25.387000+00:00  
1241      FRTCBE003954  2023-01-27  2023-08-08T07:17:25.387000+00:00  
1242      FRTCBE003959  2023-01-27  2023-08-08T07:17:25.387000+00:00

### Cohérence station - coordonnées
- 477 pdc sont associés à des stations avec plusieurs coordonnées

In [27]:
itinerance_4.loc[~itinerance_4['coordonneesXY - id_station_itinerance'], relations][200:215]

index contact_operateur          nom_enseigne          coordonneesXY  \
1507  34811  contact@e55c.com  ELECTRIC 55 CHARGING  [6.585713, 43.230239]   
1508  34812  contact@e55c.com  ELECTRIC 55 CHARGING  [6.585713, 43.230239]   
1513  34820  contact@e55c.com  ELECTRIC 55 CHARGING  [2.262841, 48.798985]   
1514  34821  contact@e55c.com  ELECTRIC 55 CHARGING  [2.262841, 48.798985]   
1515  34822  contact@e55c.com  ELECTRIC 55 CHARGING  [2.262841, 48.798985]   
1516  34823  contact@e55c.com  ELECTRIC 55 CHARGING  [2.262841, 48.798985]   
1517  34824  contact@e55c.com  ELECTRIC 55 CHARGING  [2.262841, 48.798985]   
1518  34825  contact@e55c.com  ELECTRIC 55 CHARGING  [2.262841, 48.798985]   
1519  34826  contact@e55c.com  ELECTRIC 55 CHARGING  [2.255335, 48.786899]   
1520  34827  contact@e55c.com  ELECTRIC 55 CHARGING  [2.255335, 48.786899]   
1521  34828  contact@e55c.com  ELECTRIC 55 CHARGING  [2.255335, 48.786899]   
1522  34829  contact@e55c.com  ELECTRIC 55 CHARGING  [2.255335, 48.786899]   
1523  34831  contact@e55c.com  ELECTRIC 55 CHARGING  [2.255335, 48.786899]   
1524  34832  contact@e55c.com  ELECTRIC 55 CHARGING  [2.255335, 48.786899]   
1525  34842  contact@e55c.com  ELECTRIC 55 CHARGING  [4.899412, 43.940189]   

                                   adresse_station id_station_itinerance  \
1507                   OFFICE DU TOURISME - GASSIN           FR55CP83580   
1508                   OFFICE DU TOURISME - GASSIN           FR55CP83580   
1513                       PLACE FERRARI - CLAMART           FR55CP92140   
1514                       PLACE FERRARI - CLAMART           FR55CP92140   
1515                       PLACE FERRARI - CLAMART           FR55CP92140   
1516                       PLACE FERRARI - CLAMART           FR55CP92140   
1517                       PLACE FERRARI - CLAMART           FR55CP92140   
1518                       PLACE FERRARI - CLAMART           FR55CP92140   
1519  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART           FR55CP92140   
1520  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART           FR55CP92140   
1521  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART           FR55CP92140   
1522  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART           FR55CP92140   
1523  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART           FR55CP92140   
1524  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART           FR55CP92140   
1525                  AVENUE JEAN MONNET - MORIERE           FR55CP84310   

                                       nom_station implantation_station  \
1507                   OFFICE DU TOURISME - GASSIN               Voirie   
1508                   OFFICE DU TOURISME - GASSIN               Voirie   
1513                       PLACE FERRARI - CLAMART               Voirie   
1514                       PLACE FERRARI - CLAMART               Voirie   
1515                       PLACE FERRARI - CLAMART               Voirie   
1516                       PLACE FERRARI - CLAMART               Voirie   
1517                       PLACE FERRARI - CLAMART               Voirie   
1518                       PLACE FERRARI - CLAMART               Voirie   
1519  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART               Voirie   
1520  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART               Voirie   
1521  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART               Voirie   
1522  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART               Voirie   
1523  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART               Voirie   
1524  HOPITAL BECLERE - AVENUE DE GAULLE - CLAMART               Voirie   
1525                  AVENUE JEAN MONNET - MORIERE               Voirie   

      nbre_pdc condition_acces horaires station_deux_roues  \
1507         2   Accès réservé     24/7              FALSE   
1508         2   Accès réservé     24/7              FALSE   
1513         6   Accès réservé     24/7              FALSE   
1514         6   Accès réservé     24/7              FALSE   
1515         6   Accès réserv

- exemple : une station avec plusieurs coordonnées -> incompréhension distinction station / pdc

In [28]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FR073PCAMAIEUFR', relations]

index         contact_operateur    nom_enseigne         coordonneesXY  \
1924  39670  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
1925  39672  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
1926  39673  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
1927  39674  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
1928  39675  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
1929  39676  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   
1930  39677  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   
1931  39678  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
1932  39679  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
1933  39680  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   
1934  39681  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
1935  39682  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   
1936  39683  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
1937  39684  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
1938  39685  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   
1939  39686  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   

                         adresse_station id_station_itinerance  \
1924  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1925  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1926  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1927  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1928  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1929  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1930  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1931  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1932  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1933  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1934  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1935  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1936  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1937  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1938  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1939  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   

         nom_station          implantation_station  nbre_pdc condition_acces  \
1924  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1925  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1926  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1927  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1928  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1929  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1930  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1931  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1932  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1933  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1934  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1935  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1936  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1937  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1938  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1939  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   

               horaires station_deux_roues id_pdc_itinerance    date_maj  \
1924  Mo-Su 08:00-08:00              false    FR073ETCIX5105  2022-07-21   
1925  Mo-Su 08:00-08:00   

- exemple : une station avec plusieurs coordonnées -> regroupement de stations ?

In [29]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FR55CPBP514', relations]

index contact_operateur          nom_enseigne          coordonneesXY  \
1450  34855  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242799, 49.120583]   
1451  34856  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242799, 49.120583]   
1452  34857  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242799, 49.120583]   
1453  34858  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242644, 49.120556]   
1454  34859  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242644, 49.120556]   
1455  34860  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242644, 49.120556]   

                            adresse_station id_station_itinerance  \
1450   BP CHAMPAGNE SUD - LES PETITES-LOGES           FR55CPBP514   
1451   BP CHAMPAGNE SUD - LES PETITES-LOGES           FR55CPBP514   
1452   BP CHAMPAGNE SUD - LES PETITES-LOGES           FR55CPBP514   
1453  BP CHAMPAGNE NORD - LES PETITES-LOGES           FR55CPBP514   
1454  BP CHAMPAGNE NORD - LES PETITES-LOGES           FR55CPBP514   
1455  BP CHAMPAGNE NORD - LES PETITES-LOGES           FR55CPBP514   

                                        nom_station  \
1450   BP CHAMPAGNE SUD - LES PETITES-LOGES - 69800   
1451   BP CHAMPAGNE SUD - LES PETITES-LOGES - 69800   
1452   BP CHAMPAGNE SUD - LES PETITES-LOGES - 69800   
1453  BP CHAMPAGNE NORD - LES PETITES-LOGES - 69800   
1454  BP CHAMPAGNE NORD - LES PETITES-LOGES - 69800   
1455  BP CHAMPAGNE NORD - LES PETITES-LOGES - 69800   

                     implantation_station  nbre_pdc condition_acces horaires  \
1450  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
1451  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
1452  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
1453  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
1454  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
1455  Station dédiée à la recharge rapide         3   Accès réservé     24/7   

     station_deux_roues           id_pdc_itinerance    date_maj  \
1450              FALSE  FR55CE51400GIRVATMX02AA8A1  2022-11-03   
1451              FALSE  FR55CE51400GIRVATMX02AA8A2  2022-11-03   
1452              FALSE  FR55CE51400GIRVATMX02AA8A3  2022-11-03   
1453              FALSE  FR55CE51400GIRVATMX02A6A61  2022-11-03   
1454              FALSE  FR55CE51400GIRVATMX02A6B62  2022-11-03   
1455              FALSE  FR55CE51400GIRVATMX02A6B63  2022-11-03   

                         last_modified  
1450  2023-08-08T07:16:44.947000+00:00  
1451  2023-08-08T07:16:44.947000+00:00  
1452  2023-08-08T07:16:44.947000+00:00  
1453  2023-08-08T07:16:44.947000+00:00  
1454  2023-08-08T07:16:44.947000+00:00  
1455  2023-08-08T07:16:44.947000+00:00

- exemple : quatre stations avec une localisation identique mais deux adresses différentes + id identique entre station et pdc -> erreur d'adresse et erreur d'id_station ?

In [30]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[-0.366184, 43.332508]', relations]

index contact_operateur       nom_enseigne           coordonneesXY  \
37   5066    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   
38   5076    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   
39   5077    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   
40   5079    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   

     adresse_station id_station_itinerance             nom_station  \
37  PARKING ZÉNITH 1        FRS64PMB649712  PAU - PARKING ZENITH 1   
38  PARKING ZÉNITH 1        FRS64PMB649711  PAU - PARKING ZENITH 1   
39  PARKING ZÉNITH 2        FRS64PMB649812  PAU - PARKING ZENITH 2   
40  PARKING ZÉNITH 2        FRS64PMB649811  PAU - PARKING ZENITH 2   

   implantation_station  nbre_pdc condition_acces horaires station_deux_roues  \
37               Voirie         2     Accès libre     24/7              false   
38               Voirie         2     Accès libre     24/7              false   
39               Voirie         2     Accès libre     24/7              false   
40               Voirie         2     Accès libre     24/7              false   

   id_pdc_itinerance    date_maj                     last_modified  
37    FRS64EMB649712  2023-08-09  2023-08-09T02:54:29.511000+00:00  
38    FRS64EMB649711  2023-08-09  2023-08-09T02:54:29.511000+00:00  
39    FRS64EMB649812  2023-08-09  2023-08-09T02:54:29.511000+00:00  
40    FRS64EMB649811  2023-08-09  2023-08-09T02:54:29.511000+00:00

### Incohérence 'nbre_pdc'
- 25% des pdc ont un champ 'nbre_pdc' mal documenté

- exemple : station avec un pdc 'de regroupement' qui évite de documenter plusieurs lignes

In [31]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRLUMEACACIAS11', 
               ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance           coordonneesXY  \
30412   FRLUMEACACIAS11       FRLUMEACACIAS11  [3.1485441,50.7582304]   

       nbre_pdc  
30412         4

- exemple : station avec plusieurs pdc mais l'id_station est identique à l'id_pdc

In [32]:
itinerance_init.loc[itinerance_init.coordonneesXY	 == '[2.451322, 45.66523]', 
                   ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance         coordonneesXY  nbre_pdc
5920     FRSDGEMBDG512         FRSDGPMBDG512  [2.451322, 45.66523]         2
5921     FRSDGEMBDG511         FRSDGPMBDG511  [2.451322, 45.66523]         2

- exemple : station avec plusieurs pdc mais le champs a toujours une valeur de 1

In [33]:
itinerance_init.loc[itinerance_init.coordonneesXY	 == '[-1.7548354193520386, 48.125067030488154]', 
                   ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance  \
36783    FRIENE35353A51          FRIENS35353A   
36784    FRIENE35353A11          FRIENS35353A   
36785    FRIENE35353A21          FRIENS35353A   
36786    FRIENE35353A41          FRIENS35353A   
36787    FRIENE35353A31          FRIENS35353A   

                                   coordonneesXY  nbre_pdc  
36783  [-1.7548354193520386, 48.125067030488154]         1  
36784  [-1.7548354193520386, 48.125067030488154]         1  
36785  [-1.7548354193520386, 48.125067030488154]         1  
36786  [-1.7548354193520386, 48.125067030488154]         1  
36787  [-1.7548354193520386, 48.125067030488154]         1

### cohérence station - enseigne
- exemple de station avec plusieurs noms d'enseigne

In [34]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRS27PBARREOUCHEMOMORT', relations]

index            contact_operateur nom_enseigne         coordonneesXY  \
36185  36643  pascal.lhermitte@siege27.fr     SIEGE266  [0.66299, 48.946159]   
36187  36645  pascal.lhermitte@siege27.fr     SIEGE265  [0.66299, 48.946159]   

                                      adresse_station   id_station_itinerance  \
36185   Place du monument aux Morts La Barre en Ouche  FRS27PBARREOUCHEMOMORT   
36187   Place du monument aux Morts La Barre en Ouche  FRS27PBARREOUCHEMOMORT   

      nom_station implantation_station  nbre_pdc condition_acces horaires  \
36185      900120               Voirie         2     Accès libre     24/7   
36187      900120               Voirie         2     Accès libre     24/7   

      station_deux_roues        id_pdc_itinerance    date_maj  \
36185              false  FRS27EBARREOUCHEMOMORTG  2021-06-15   
36187              false  FRS27EBARREOUCHEMOMORTD  2021-06-15   

                          last_modified  
36185  2023-08-08T07:16:29.698000+00:00  
36187  2023-08-08T07:16:29.698000+00:00

### cohérence station - horaires
- exemple de station avec plusieurs types d'horaire

In [35]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRS23D2302001', relations]

index  contact_operateur               nom_enseigne  \
36690  37162  contact@mobive.fr  Réseau de recharge Creuse   
36691  37163  contact@mobive.fr  Réseau de recharge Creuse   

                                coordonneesXY adresse_station  \
36690  [2.297630254367735, 45.98220366857867]  Place du Treix   
36691  [2.297630254367735, 45.98220366857867]  Place du Treix   

      id_station_itinerance           nom_station implantation_station  \
36690         FRS23D2302001  BELLEGARDE EN MARCHE       Parking public   
36691         FRS23D2302001  BELLEGARDE EN MARCHE       Parking public   

       nbre_pdc condition_acces           horaires station_deux_roues  \
36690         2     Accès libre  Mo-Su 00:00-00:01              False   
36691         2     Accès libre  Mo-Su 00:00-00:02              False   

      id_pdc_itinerance    date_maj                     last_modified  
36690   FRS23P230200101  2023-03-09  2023-08-08T07:16:16.268000+00:00  
36691   FRS23P230200102  2023-03-09  2023-08-08T07:16:16.268000+00:00

### cohérence station - deux-roues
- exemple de station avec plusieurs indications d'accessibilité deux-roues

In [36]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRV75PPX1902', relations]

index                        contact_operateur nom_enseigne  \
27051  27473  supervision-ev.france@totalenergies.com       Belib'   
27071  27493  supervision-ev.france@totalenergies.com       Belib'   
27072  27494  supervision-ev.france@totalenergies.com       Belib'   
27073  27495  supervision-ev.france@totalenergies.com       Belib'   
27074  27496  supervision-ev.france@totalenergies.com       Belib'   
27092  27514  supervision-ev.france@totalenergies.com       Belib'   
27093  27515  supervision-ev.france@totalenergies.com       Belib'   

                coordonneesXY                    adresse_station  \
27051  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
27071  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
27072  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
27073  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
27074  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
27092  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
27093  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    

      id_station_itinerance                   nom_station  \
27051          FRV75PPX1902  Paris | Avenue de Flandre 61   
27071          FRV75PPX1902  Paris | Avenue de Flandre 61   
27072          FRV75PPX1902  Paris | Avenue de Flandre 61   
27073          FRV75PPX1902  Paris | Avenue de Flandre 61   
27074          FRV75PPX1902  Paris | Avenue de Flandre 61   
27092          FRV75PPX1902  Paris | Avenue de Flandre 61   
27093          FRV75PPX1902  Paris | Avenue de Flandre 61   

      implantation_station  nbre_pdc condition_acces horaires  \
27051               Voirie         7     Accès libre     24/7   
27071               Voirie         7     Accès libre     24/7   
27072               Voirie         7     Accès libre     24/7   
27073               Voirie         7     Accès libre     24/7   
27074               Voirie         7     Accès libre     24/7   
27092               Voirie         7     Accès libre     24/7   
27093               Voirie         7     Accès libre     24/7   

      station_deux_roues id_pdc_itinerance    date_maj  \
27051              False     FRV75EPX19027  2023-07-06   
27071              False     FRV75EPX19026  2023-07-06   
27072              False     FRV75EPX19025  2023-07-06   
27073              False     FRV75EPX19024  2023-07-06   
27074               True     FRV75EPX19023  2023-07-06   
27092              False     FRV75EPX19021  2023-07-06   
27093               True     FRV75EPX19022  2023-07-06   

                          last_modified  
27051  2023-08-08T07:17:25.957000+00:00  
27071  2023-08-08T07:17:25.957000+00:00  
27072  2023-08-08T07:17:25.957000+00:00  
27073  2023-08-08T07:17:25.957000+00:00  
27074  2023-08-08T07:17:25.957000+00:00  
27092  2023-08-08T07:17:25.957000+00:00  
27093  2023-08-08T07:17:25.957000+00:00

### cohérence coordonnées - adresse
- exemple de plusieurs stations avec des coordonnées identiques mais des adresses différentes

In [37]:
itinerance_5.loc[itinerance_5.coordonneesXY == '[2.460441, 50.78763]', relations]

Empty DataFrame
Columns: [index, contact_operateur, nom_enseigne, coordonneesXY, adresse_station, id_station_itinerance, nom_station, implantation_station, nbre_pdc, condition_acces, horaires, station_deux_roues, id_pdc_itinerance, date_maj, last_modified]
Index: []

### données d'un opérateur
- exemple Virta

In [38]:
itinerance_operateur = itinerance_5_full.loc[itinerance_5_full.contact_operateur == 'support@virta.global']
#itinerance_operateur.loc[itinerance_operateur['adresse_station - coordonneesXY'] == False, relations]
itinerance_operateur.loc[itinerance_operateur['nom_station - id_station_itinerance'] == False, relations]

Empty DataFrame
Columns: [index, contact_operateur, nom_enseigne, coordonneesXY, adresse_station, id_station_itinerance, nom_station, implantation_station, nbre_pdc, condition_acces, horaires, station_deux_roues, id_pdc_itinerance, date_maj, last_modified]
Index: []